brew_view/static/src/js/controllers/admin_system.js

```diff
@@ -166,5 +140,95 @@ export default function systemAdminController(
     $scope.systems.errorMessage = response.data.message;
   };
 
-  SystemService.getSystems().then($scope.successCallback, $scope.failureCallback);
+  let socketError = false;
+
+  function websocketConnect() {
+    if (window.WebSocket && !socketError) {
+      let proto = window.location.protocol === 'https:' ? 'wss://' : 'ws://';
+      let event_url = proto + window.location.host + '/api/v1/socket/events';
+
+      let socketConnection = $websocket(event_url);
+      socketConnection.onMessage(handleWebsocketMessage);
+
+      // If the connection is broken attempt to reconnect.
+      // If this is caused by brew-view stopping the reconnect attempt will
+      // probably error. This isn't ideal, but we can't distinguish between a
+      // 'real' error and brew-view being down, so we live with this for now.
+      socketConnection.onClose(websocketConnect);
+      socketConnection.onError(function() {
+        socketError = true;
+      });
+
+      $scope.$on('destroy', function() {
+        if (angular.isDefined(socketConnection)) {
+          socketConnection.close();
+          socketConnection = undefined;
+        }
+      });
+    }
+  }
+
+  function handleWebsocketMessage(message) {
+    let event = JSON.parse(message.data);
+    console.log(event);
+
+    switch(event.name) {
+      case 'INSTANCE_INITIALIZED':
+        updateInstanceStatus(event.payload.id, 'RUNNING');
+        break;
+      case 'INSTANCE_STOPPED':
+        updateInstanceStatus(event.payload.id, 'STOPPED');
+        break;
+      case 'SYSTEM_REMOVED':
+        removeSystem(event.payload.id);
+        break;
+    }
+  }
+
+  function updateInstanceStatus(id, newStatus) {
+    if (newStatus === undefined) return;
+
+    for (let system_name in $scope.systems.data) {
```

Comment:
Shouldn't this also be `of` instead of `in`?

Reply:
No, that `groupBy` call in the success callback means that `$scope.systems.data` is an object of system name -> list of systems. So the outer loop is 'in' because it's an object, and the inner loop is 'of' because it's a list.

Reply:
:+1: I've been using `for ... of` for all my new JS writing. It seems safer, but I agree, this will work!